### PR TITLE
EE-7359: Add clarification regarding the usage of access tokens in the connection creation.

### PR DESCRIPTION
**Pull request by ndinu**
…e connection creation.

### CODE REVIEW

File: modules/ROOT/pages/define-configurations-and-connection-providers.adoc [hunk 1]
```diff
@@ -85,6 +85,44 @@ Therefore, the getAccessToken() method *must* be invoked on the state object for
 
 This consideration is especially important when using third-party client libraries that might perform caching.
 
+=== Don't use the access token in the connection creation
```
Comment (sarathecoubian) [inline modules/ROOT/pages/define-configurations-and-connection-providers.adoc]:
```suggestion
=== Do Not Use the Access Token in the Connection Creation
```

File: modules/ROOT/pages/define-configurations-and-connection-providers.adoc [hunk 2]
```diff
@@ -85,6 +85,44 @@ Therefore, the getAccessToken() method *must* be invoked on the state object for
...
+  private String getUserId(String accessToken){
+    ...
+  }
+
+  ...
+
+}
+----
+
+Now, the first time the LazyValue is used, the information will be retrieved, and all later usages will get the already resolved value.
```
Comment (sarathecoubian) [inline modules/ROOT/pages/define-configurations-and-connection-providers.adoc]:
```suggestion
As such, the first time `LazyValue` is used, the information is retrieved, and all later usages get the value that has already been resolved.
```

File: modules/ROOT/pages/define-configurations-and-connection-providers.adoc [hunk 3]
```diff
@@ -85,6 +85,44 @@ Therefore, the getAccessToken() method *must* be invoked on the state object for
 
 This consideration is especially important when using third-party client libraries that might perform caching.
 
+=== Don't use the access token in the connection creation
+
+The creation of the connection object in the ConnectionProvider#connect method must not depend on the presence of a valid access token. Even if the connection needs to use the token to perform a one-time only request, that should not be done in the creation of the connection. The request must be deferred to when the connection is first used.
+
+To this end, the org.mule.runtime.api.util.LazyValue class can be leveraged. So that when a value is first used, the action of retrieving the value is done, and later uses will get the value that was cached from the first time.
+
+Here is an example of how this can be done:
```
Comment (sarathecoubian) [inline modules/ROOT/pages/define-configurations-and-connection-providers.adoc]:
```suggestion
The following example shows how to use the `org.mule.runtime.api.util.LazyValue` class:
```

File: modules/ROOT/pages/define-configurations-and-connection-providers.adoc [hunk 4]
```diff
@@ -85,6 +85,44 @@ Therefore, the getAccessToken() method *must* be invoked on the state object for
 
 This consideration is especially important when using third-party client libraries that might perform caching.
 
+=== Don't use the access token in the connection creation
+
+The creation of the connection object in the ConnectionProvider#connect method must not depend on the presence of a valid access token. Even if the connection needs to use the token to perform a one-time only request, that should not be done in the creation of the connection. The request must be deferred to when the connection is first used.
```
Comment (sarathecoubian) [inline modules/ROOT/pages/define-configurations-and-connection-providers.adoc]:
```suggestion
The creation of the connection object in the `ConnectionProvider#connect` method must not depend on the presence of a valid access token. Connections that need to use the token to perform a one-time-only request should not rely on the creation of the connection. The request must be deferred to when the connection is first used.
```

File: modules/ROOT/pages/define-configurations-and-connection-providers.adoc [hunk 5]
```diff
@@ -85,6 +85,44 @@ Therefore, the getAccessToken() method *must* be invoked on the state object for
...
+
+To this end, the org.mule.runtime.api.util.LazyValue class can be leveraged. So that when a value is first used, the action of retrieving the value is done, and later uses will get the value that was cached from the first time.
```
Comment (sarathecoubian) [inline modules/ROOT/pages/define-configurations-and-connection-providers.adoc]:
```suggestion
To adhere to this practice, you can use the `org.mule.runtime.api.util.LazyValue` class. When a value is initially used, the value is retrieved; subsequent uses get the value that was cached initially.
```

Comment (sarathecoubian) [reply modules/ROOT/pages/define-configurations-and-connection-providers.adoc]: I do not really understand which "first time" this references, so if we need to modify this further, please let me know.

**Review by sarathecoubian — CHANGES_REQUESTED** [review]

File: modules/ROOT/pages/define-configurations-and-connection-providers.adoc (hunk 5 above)
Comment (ndinu) [reply modules/ROOT/pages/define-configurations-and-connection-providers.adoc]: The "first time" references to the moment when the value is first used. You create something that knows how to resolve a value, when you use it for the first time the value is resolved, later usages will return the value that was given before without executing the resolution again.

Comment (sarathecoubian) [reply modules/ROOT/pages/define-configurations-and-connection-providers.adoc]: To adhere to this practice, you can use the `org.mule.runtime.api.util.LazyValue` class. After the value is retrieved for the first time, subsequent uses get the same value that was cached in the initial retrieval.

**Review by sarathecoubian — APPROVED** [review]

**Pull request merged**